drivers/display/display_tn0xxx.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2023 Alen Daniel
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT kyo_tn0xxx
+
+#include <zephyr/logging/log.h>
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/display.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/logging/log.h>
+#include "lvgl.h"
+#include "lvgl_display.h"
+
+/* Driver for Kyocera's 2.16" RESOLUTION MEMORY IN PIXEL (MIP) TFT (TN0216ANVNANN)
+ * Note:
+ * high/1 means white, low/0 means black
+ * SPI interface expects LSB first
+ * see more notes in boards/shields/tn0xxx/doc/index.rst
+ */
+LOG_MODULE_REGISTER(tn0xxx, CONFIG_DISPLAY_LOG_LEVEL);
+
+#define TN0XXX_PANEL_WIDTH  DT_INST_PROP(0, width)
+#define TN0XXX_PANEL_HEIGHT DT_INST_PROP(0, height)
+
+#define TN0XXX_PIXELS_PER_BYTE 8
+
+#define LCD_ADDRESS_LEN_BITS          8
+#define LCD_DUMMY_SPI_CYCLES_LEN_BITS 32
+#define DUMMY_BYTE                    0x00
+
+/* Data packet format
+ * +-------------------+-------------------+--------------------+
+ * | line addr (8 bits) | data (8 WIDTH bits) | dummy (32 bits) |
+ * +-------------------+-------------------+--------------------+
+ */
+
+struct tn0xxx_config_s {
+	struct spi_dt_spec bus;
+};
+
+struct tn0xxx_data_s {
+	enum display_orientation orientation;
+	enum display_pixel_format pixel_format;
+};
+
+static int tn0xxx_blanking_off(const struct device *dev)
+{
+	LOG_WRN("Unsupported");
+	return -ENOTSUP;
+}
+
+static int tn0xxx_blanking_on(const struct device *dev)
+{
+	LOG_WRN("Unsupported");
+	return -ENOTSUP;
+}
+
+static int tn0xxx_read(const struct device *dev, const uint16_t x, const uint16_t y,
+		       const struct display_buffer_descriptor *desc, void *buf)
+{
+	LOG_ERR("not supported");
+	return -ENOTSUP;
+}
+
+static void *tn0xxx_get_framebuffer(const struct device *dev)
+{
+	LOG_ERR("not supported");
+	return NULL;
+}
+
+static int tn0xxx_set_brightness(const struct device *dev, const uint8_t brightness)
+{
+	LOG_WRN("not supported");
+	return -ENOTSUP;
+}
+
+static int tn0xxx_set_contrast(const struct device *dev, uint8_t contrast)
+{
+	LOG_WRN("not supported");
+	return -ENOTSUP;
+}
+
+static int tn0xxx_set_orientation(const struct device *dev,
+				  const enum display_orientation new_orientation)
+{
+	LOG_WRN("not supported");
+	return -ENOTSUP;
+}
+
+static int tn0xxx_set_pixel_format(const struct device *dev, const enum display_pixel_format pf)
+{
+
+	lv_disp_t *disp = lv_disp_get_default();
+	struct lvgl_disp_data *disp_data = disp->driver->user_data;
+	struct display_capabilities *caps = &disp_data->cap;
+	struct tn0xxx_data_s *data = dev->data;
+
+	if (pf & caps->supported_pixel_formats) {
+		caps->current_pixel_format = pf;
+		data->pixel_format = pf;
+		lv_disp_drv_update(disp, disp->driver);
+		return 0;
```

Comment:
The driver may only use the API provided by https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/display.h. The driver may not call any gui library or subsystem, and certainly not any third party library or subsystem.

Reply:
Thank you for this feedback, good points. I have updated the commit to reflect this